third_party/bigframes_vendored/pandas/core/computation/eval.py

```diff
@@ -0,0 +1,401 @@
+        level (int, optional):
+            The number of prior stack frames to traverse and add to the current
+            scope. Most users will **not** need to change this parameter.
+        target (object, optional, default None):
+            This is the target object for assignment. It is used when there is
+            variable assignment in the expression. If so, then `target` must
+            support item assignment with string keys, and if a copy is being
+            returned, it must also support `.copy()`.
+        inplace (bool, default False):
+            If `target` is provided, and the expression mutates `target`, whether
+            to modify `target` inplace. Otherwise, return a copy of `target` with
+            the mutation.
+
+    Returns:
+        ndarray, numeric scalar, DataFrame, Series, or None:
+        The completion value of evaluating the given code or None if ``inplace=True``.
+
+    Raises:
+        ValueError:
+            There are many instances where such an error can be raised:
+
+            - `target=None`, but the expression is multiline.
+            - The expression is multiline, but not all them have item assignment.
+            An example of such an arrangement is this:
+
+            a = b + 1
+            a + 2
+
+            Here, there are expressions on different lines, making it multiline,
+            but the last line has no variable assigned to the output of `a + 2`.
+            - `inplace=True`, but the expression is missing item assignment.
+            - Item assignment is provided, but the `target` does not support
+            string item assignment.
+            - Item assignment is provided and `inplace=False`, but the `target`
+            does not support the `.copy()` method
+
+    """
+    inplace = validate_bool_kwarg(inplace, "inplace")
+
+    exprs: list[str | BinOp]
+    if isinstance(expr, str):
+        _check_expression(expr)
+        exprs = [e.strip() for e in expr.splitlines() if e.strip() != ""]
+    else:
+        # ops.BinOp; for internal compat, not intended to be passed by users
+        exprs = [expr]
+    multi_line = len(exprs) > 1
+
+    if multi_line and target is None:
+        raise ValueError(
+            "multi-line expressions are only valid in the "
+            "context of data, use DataFrame.eval"
+        )
+    engine = _check_engine(engine)
+    _check_parser(parser)
+    _check_resolvers(resolvers)
+
+    ret = None
+    first_expr = True
+    target_modified = False
+
+    for expr in exprs:
+        expr = _convert_expression(expr)
+        _check_for_locals(expr, level, parser)
+
+        # get our (possibly passed-in) scope
+        env = ensure_scope(
+            level + 1,
+            global_dict=global_dict,
+            local_dict=local_dict,
+            resolvers=resolvers,
+            target=target,
+        )
+
+        parsed_expr = Expr(expr, engine=engine, parser=parser, env=env)
+
+        if engine == "numexpr" and (
```

Comment:
I'm wondering if we want to catch this and raise immediately?

Aside: I could imagine adding other "engines" in future. My initial thought for this issue was for a "sql" engine to materialize to a table and/or allow injecting arbitrary SQL, but that gets tricky since our labels don't always match the column IDs.

Reply:
Went ahead and remove remaining references to "numexpr". Attempts to use this engine will result in exception.